src/context.cc

```diff
@@ -90,44 +90,36 @@ void Context::setVariables(const EvalContext *evalctx, const AssignmentList &arg
 	}
 }
 
-void Context::set_variable(const std::string &name, const ValuePtr &value)
+// sink for value takes &&
+void Context::set_variable(const std::string &name, Value&& value)
 {
-	if (is_config_variable(name)) this->config_variables[name] = value;
-	else this->variables[name] = value;
+	if (is_config_variable(name)) this->config_variables[name] = std::move(value);
+	else this->variables[name] = std::move(value);
 }
 
-void Context::set_variable(const std::string &name, const Value &value)
-{
-	set_variable(name, ValuePtr(value));
-}
-
-void Context::set_constant(const std::string &name, const ValuePtr &value)
+// sink for value takes &&
+void Context::set_constant(const std::string &name, Value&& value)
 {
 	if (this->constants.find(name) != this->constants.end()) {
 		PRINTB("WARNING: Attempt to modify constant '%s'.", name);
 	}
 	else {
-		this->constants[name] = value;
+		this->constants[name] = std::move(value);
 	}
 }
 
-void Context::set_constant(const std::string &name, const Value &value)
-{
-	set_constant(name, ValuePtr(value));
-}
-
-void Context::apply_variables(const Context &other)
+void Context::take_variables(Context &other)
 {
-	for (const auto &var : other.variables) {
-		set_variable(var.first, var.second);
+	for (auto &var : other.variables) {
+		set_variable(var.first, std::move(var.second));
```

Comment:
This `take_variables` function was an optimization idea I had to avoid copies when and move variables from a temporary context into some other context, but I think t-paul replaced calls to this with the more traditional `apply_variables`, when he merged function-literals in value-rework-2.  
I'm guessing it may have possibly caused issues with closures keeping a context alive even if it was "temporary", and probably best to just not move things out of containers which requires too much hard thinking about cases like that.
Since it was no longer being called I just went ahead and removed the definition in #3102.